src/xrpld/app/tx/detail/PermissionedDomainSet.cpp

```diff
@@ -0,0 +1,158 @@
+}
+
+TER
+PermissionedDomainSet::preclaim(PreclaimContext const& ctx)
+{
+    if (!ctx.view.read(keylet::account(ctx.tx.getAccountID(sfAccount))))
+        return tefINTERNAL;
+
+    assert(ctx.tx.isFieldPresent(sfAcceptedCredentials));
+    auto const credentials = ctx.tx.getFieldArray(sfAcceptedCredentials);
+    for (auto const& credential : credentials)
+    {
+        if (!credential.isFieldPresent(sfIssuer) ||
+            !credential.isFieldPresent(sfCredentialType))
+        {
+            return temMALFORMED;
+        }
+        if (credential.getFieldVL(sfCredentialType).empty())
+            return temMALFORMED;
+        if (!ctx.view.read(keylet::account(credential.getAccountID(sfIssuer))))
+            return temBAD_ISSUER;
+    }
+
+    if (!ctx.tx.isFieldPresent(sfDomainID))
+        return tesSUCCESS;
+    auto const domain = ctx.tx.getFieldH256(sfDomainID);
+    if (domain == beast::zero)
+        return temMALFORMED;
+    auto const sleDomain = ctx.view.read({ltPERMISSIONED_DOMAIN, domain});
+    if (!sleDomain)
+        return tecNO_ENTRY;
+    auto const owner = sleDomain->getAccountID(sfOwner);
+    auto account = ctx.tx.getAccountID(sfAccount);
+    if (owner != account)
+        return temINVALID_ACCOUNT_ID;
+
+    return tesSUCCESS;
+}
+
+/** Attempt to create the Permissioned Domain. */
+TER
+PermissionedDomainSet::doApply()
+{
+    auto const ownerSle = view().peek(keylet::account(account_));
+
+    // All checks have already been done. Just update credentials. Same logic
+    // for either new domain or updating existing.
+    // Silently remove duplicates.
+    auto updateSle = [this](std::shared_ptr<STLedgerEntry> const& sle) {
+        auto credentials = ctx_.tx.getFieldArray(sfAcceptedCredentials);
+        std::map<uint256, STObject> hashed;
+        for (auto const& c : credentials)
+            hashed.insert({c.getHash(HashPrefix::transactionID), c});
+        if (credentials.size() > hashed.size())
+        {
+            credentials = STArray();
+            for (auto const& e : hashed)
+                credentials.push_back(e.second);
+        }
+
+        credentials.sort(
+            [](STObject const& left, STObject const& right) -> bool {
+                if (left.getAccountID(sfIssuer) < right.getAccountID(sfIssuer))
+                    return true;
+                if (left.getAccountID(sfIssuer) == right.getAccountID(sfIssuer))
+                {
+                    if (left.getFieldVL(sfCredentialType) <
+                        right.getFieldVL(sfCredentialType))
+                    {
+                        return true;
+                    }
+                }
+                return false;
+            });
+        sle->setFieldArray(sfAcceptedCredentials, credentials);
+    };
+
+    if (ctx_.tx.isFieldPresent(sfDomainID))
+    {
+        // Modify existing permissioned domain.
+        auto sleUpdate = view().peek(
+            {ltPERMISSIONED_DOMAIN, ctx_.tx.getFieldH256(sfDomainID)});
```

Comment:
I would add `assert(sleUpdate)`, but that's just me.

Reply:
fixed